paasta_tools/cli/cmds/exec.py

```diff
@@ -0,0 +1,143 @@
+        "-s",
+        "--service",
+        help="The name of the service you wish to inspect. Defaults to autodetect.",
+    ).completer = lazy_choices_completer(list_services)
+    exec_parser.add_argument(
+        "-c",
+        "--cluster",
+        help="The cluster to exec into.",
+        type=str,
+    ).completer = completer_clusters
+    exec_parser.add_argument(
+        "-i",
+        "--instance",
+        help="The instance to exec into.",
+        type=str,
+    ).completer = completer_clusters
+    pod_help = "The pod to exec into. Defaults to a random running pod for the service."
+    exec_parser.add_argument("-p", "--pod", help=pod_help)
+
+    exec_parser.add_argument(
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help=f"Define a different soa config directory. Defaults to %(default)s.",
+    )
+
+    exec_parser.set_defaults(command=paasta_exec)
+
+
+def completer_clusters(prefix, parsed_args, **kwargs):
+    service = parsed_args.service or guess_service_name()
+    if service in list_services():
+        return list_clusters(service)
+    else:
+        return list_clusters()
+
+
+def paasta_exec(args: argparse.Namespace) -> int:
+    """Execs into a running Paasta service.
+    :param args: argparse.Namespace obj created from sys.args by cli"""
+    soa_dir = args.soa_dir
+
+    service = figure_out_service_name(args, soa_dir)
+
+    if (
+        args.cluster is None
+        or args.instance is None
+        or len(args.instance.split(",")) > 2
+    ):
+        print(
+            PaastaColors.red("You must specify one cluster and one instance."),
+            file=sys.stderr,
+        )
+        return 1
+    cluster = args.cluster
+    if verify_instances(args.instance, service, [cluster]):
+        return 1
+
+    instance = args.instance
+    pods = (
+        subprocess.run(
+            [
+                f"kubectl-{cluster}",
+                "get",
+                "po",
+                "-n",
+                "paasta",
+                "-l",
+                f"paasta.yelp.com/service={service},paasta.yelp.com/instance={instance}",
+                "--field-selector=status.phase==Running",
+                "--no-headers",
+                "-o",
+                "custom-columns=:.metadata.name",
+                "|",
+                "shuf",
+            ],
+            stdout=subprocess.PIPE,
+        )
+        .stdout.decode("utf-8")
+        .strip()
+        .split("\n")
+    )
+    if args.pod is None:
+        pod = random.choice(pods)
+    else:
+        pod = args.pods
+        if pod not in pods:
+            pod = ""
+    if not pod:
+        print(
+            PaastaColors.red("We can't find a running pod for your service."),
+            file=sys.stderr,
+        )
+        return 1
+
+    cmd = f"kubectl-{cluster} -i -t -n paasta exec {pod} -- sh -c '\"clear; (bash || ash || sh)\"'"
```

Comment:
same as above, imo this should use the API and if not it shouldn't use the wrapper

Reply:
imo, we should also just default to running a shell that's in our base images - i'd rather not encourage folks to start adding every shell under the soon to their images on the off chance that they need to exec in to debug

Reply:
for future improvements (unless you want extra credit): would be nice to allow users to customize the command here

Reply:
@nemacysts the clear is for giving the user a blank terminal when they exec into something. Also I do agree about the shell I'm switching it to default to bash -> ash -> sh (in that order) or let the user run something arbitrarily.

Reply:
we should just pick one imo (and probably bash since that’s the default folks get on a devbox) - having this fallback code doesn’t really add much since i’d be surprised if we ever end up actually using any of the fallback options

Reply:
(i'd also encourage us to not mess with people's terminals and remove the clear - i'd personally be annoyed if kubectl exec cleared my screen, so i wouldn’t want paasta to do that either)